### ISSUE TITLE
Installing logsfilter causes bootloop on art

### ISSUE THREAD

**Issue by mittorn** [pylerSM/LogsFilter#1]
Full art crash log:
http://pastebin.com/f9818Lzk

**Comment by pylerSM** [comment]
Call @rovo89 for help.

**Comment by rovo89** [comment]
@pylerSM No, please check the log in detail before you point at others. This issue is caused by your module. You have hooked Log.w(), and in your callback, you call XSharedPreferences.reload(), which calls Log.w() again. That results in infinite recursion, which is ended by an overflow (usually a stack overflow, but in this case, the local references table has an overflow).

I strongly recommend not to hook such central methods. If you must do it, keep your callback as fast as possible. Triggering a reload() every time anything is logged is totally crazy, as this is an expensive call that involves IPC, file access, locking etc.

As a solution, skip the reload and ask users to reboot everytime they change their settings. That shouldn't happen too often anyway. Then you also don't need to check whether your module is enabled every time and you can cache all the variables instead of reading them from the preferences every time.

**Comment by pylerSM** [comment]
Good solution. Thanks.

**Comment by pylerSM** [comment]
But what is crazy, it doesn't crash on Dalvik.

**Comment by rovo89** [comment]
That's not because of Dalvik or ART, it's because of the hoops I had to jump through in order to work around SELinux. On older Android versions, the check whether the file has changed can be rather simple, basically just a normal stat(). Also, the file could be read directly. On Lollipop, I have to go through complex native services to make this work. It's more likely that somewhere on that way, a warning is generated.
And just because it works somehow, it doesn't mean it's good...

**Comment by pylerSM** [comment]
Yep, sure, I was just interested in. I will fix this issue hopefully tomorrow. 

// now some OT for @rovo89  (I don't want to mess Xposed github issues section with my questions so let's start)

Today I have just started project to provide simple solution for Xposed installation (imho, so many Lollipop buils + MR1 + Samsung so this tool could solve it). Simply I do some checks in app and pick correct build. If there are new builds, I don't have to update app, since we get links from json file https://github.com/pylerSM/XposedManager/blob/master/xposed_builds.json, I/anybody can update links. Code is WIP, but detecting correctfiles should work now. 

TODO: Support for Samsung Xposed builds, downloading files, installation.

If I can ask you, do you plan merge Xposed Installer 2.7 into 3.0 alpha (so one universal app) or builds will stay separated? 

I suppose you are not going to provide "root-only, w/o need of custom recovery" installation method for Xposed Lollipop, am I right? It is impossible to replace art files on the fly, right?

**Comment by rovo89** [comment]
> If I can ask you, do you plan merge Xposed Installer 2.7 into 3.0 alpha (so one universal app) or builds will stay separated?

I hope to have one universal app again in the future, which can then hopefully also install the framework on the fly. No idea when this will happen though.
